File: engine/do.mjs
// rabadon do — the motor. Task in, verified output out.
//
// The failure this kills, in the builder's words: a 10-step pipeline hits a
// fault at one gate, the flow says "continue" anyway, and the final output is
// ECLECTIC — half old plan, half improvisation, coherent to nobody. rabadon's
// job is to foresee that moment AT THE GATE WHERE IT IS BORN and pivot there,
// in the background, so the only thing the builder ever sees is the output
// they asked for.
//
// Architecture (token-frugal by construction):
//   PLAN    one LLM call. The task is decomposed into steps; EVERY step gets a
//           MECHANICAL contract (command exits green / file exists / pattern
//           present). No LLM judges a contract — checks are free forever.
//   RUN     deterministic runner. cmd steps cost zero tokens; work steps are
//           the actual labor (headless claude, narrow instruction).
//   GATE    after every step its contract runs BEFORE anything flows onward.
//           A broken contract cannot leak downstream — the eclectic output is
//           structurally impossible, not hopefully avoided.
//   PIVOT   on a broken gate: one repair attempt (LLM, given the exact
//           contract failure). Still broken: ONE replan of the remaining
//           steps from the point of deviation. Bounded; then fail-closed.
//   ACCEPT  the task-level contract runs at the end. "Done" is a measurement.
//
// Remove the LLM and what remains: the runner, the contract engine, the gate
// semantics, the bounds, the event stream. That is the moat test passing.

import fs from 'node:fs';
import path from 'node:path';
import { execSync, execFile } from 'node:child_process';
import { emitter } from '../core/bus.mjs';

const PLAN_PROMPT = (task, dir, tree) => `You are rabadon's planner. Decompose the task into a pipeline whose every step is verifiable BY MACHINE. Return ONLY JSON, no fences:

{ "steps": [ { "id": "s1", "kind": "cmd" | "work",
    "do": "<kind=cmd: the literal shell command. kind=work: a narrow, self-contained instruction for a coding agent>",
    "contract": [ { "type": "cmd", "run": "<shell command>", "passPattern": "<regex that appears ONLY on success; omit to accept exit 0>" }
                | { "type": "fileExists", "path": "<relative path>" }
                | { "type": "fileContains", "path": "<relative path>", "pattern": "<regex>" } ] } ],
  "accept": [ <same contract shapes — the TASK-level definition of done> ] }

Laws:
- 3 to 10 steps. Prefer kind=cmd (zero cost) wherever a shell command can do the job; kind=work only where judgment/writing code is unavoidable.
- EVERY step gets at least one contract, and contracts must be mechanical — a human or model must never be needed to decide pass/fail.
- accept[] must capture what the user actually asked for, not a proxy.
- Steps run in the project directory: ${dir}
- Project files (top level): ${tree}

THE TASK: ${task}`;

function runClaude(prompt, { cwd, timeoutMs = 20 * 60000, worker = false } = {}) {
  return new Promise((resolve, reject) => {
    const args = ['-p', '--output-format', 'text'];
    if (worker) { args.push('--permission-mode', 'acceptEdits', '--allowedTools', 'Read,Edit,Write,Bash,Grep,Glob'); }
    const child = execFile('claude', args, { cwd, timeout: timeoutMs, maxBuffer: 8 * 1024 * 1024 },
      (err, stdout, stderr) => err ? reject(new Error((stderr || err.message).slice(0, 300))) : resolve(String(stdout).trim()));
    child.stdin.write(prompt);
    child.stdin.end();
  });
}

function checkContract(contract, dir) {
  const fails = [];
  for (const c of contract || []) {
    try {
      if (c.type === 'cmd') {
        let out = '';
        try { out = execSync(c.run, { cwd: dir, encoding: 'utf8', timeout: 10 * 60000, stdio: ['ignore', 'pipe', 'pipe'], maxBuffer: 16 * 1024 * 1024 }); }
        catch (e) { fails.push({ check: c.run, why: `exit!=0: ${String((e.stdout || '') + (e.stderr || '')).slice(-300)}` }); continue; }
        if (c.passPattern && !new RegExp(c.passPattern, 'im').test(out)) fails.push({ check: c.run, why: `pass pattern "${c.passPattern}" absent. tail: ${out.slice(-200)}` });
      } else if (c.type === 'fileExists') {
        if (!fs.existsSync(path.join(dir, c.path))) fails.push({ check: `fileExists ${c.path}`, why: 'missing' });
      } else if (c.type === 'fileContains') {
        const body = fs.readFileSync(path.join(dir, c.path), 'utf8');
        if (!new RegExp(c.pattern, 'im').test(body)) fails.push({ check: `fileContains ${c.path}`, why: `pattern "${c.pattern}" absent` });
      } else {
        fails.push({ check: JSON.stringify(c).slice(0, 60), why: 'unknown contract type (fail-closed)' });
      }
    } catch (e) {
      fails.push({ check: JSON.stringify(c).slice(0, 60), why: String(e.message).slice(0, 120) });
    }
  }
  return fails;
}

const parseJson = (t) => { try { return JSON.parse(t.replace(/^```(?:json)?\s*|\s*```$/g, '')); } catch { return null; } };

export async function runDo(task, dir, { maxReplans = 2 } = {}) {
  const emit = emitter({ pipe: `${path.basename(dir)}:do` });
  const log = (m) => process.stdout.write(m + '\n');
  const tree = fs.readdirSync(dir).filter((f) => !f.startsWith('.')).slice(0, 40).join(', ');

  log(`plan: deriving the pipeline from the task (one model call)…`);
  let plan = parseJson(await runClaude(PLAN_PROMPT(task, dir, tree), { cwd: dir, timeoutMs: 5 * 60000 }));
  if (!plan || !Array.isArray(plan.steps) || !plan.steps.length) throw new Error('rabadon do: planner returned no usable pipeline');
  fs.mkdirSync(path.join(dir, '.rabadon'), { recursive: true });
  fs.writeFileSync(path.join(dir, '.rabadon', 'do-plan.json'), JSON.stringify({ task, ...plan }, null, 2));
  emit('RUN_START', { steps: plan.steps.map((s) => s.id), bound: { maxRepairsPerStep: 1, maxReplans } });
  log(`plan: ${plan.steps.length} steps [${plan.steps.map((s) => `${s.id}:${s.kind}`).join(' ')}] — written to .rabadon/do-plan.json`);

  let replans = 0;
  let queue = [...plan.steps];
  const done = [];

  while (queue.length) {
    const step = queue.shift();
    emit('STEP_START', { step: step.id });
    log(`${step.id} (${step.kind}): ${String(step.do).slice(0, 90)}`);

    const execStep = async (instr) => {
      if (step.kind === 'cmd') {
        try { execSync(instr, { cwd: dir, encoding: 'utf8', timeout: 15 * 60000, stdio: ['ignore', 'pipe', 'pipe'], maxBuffer: 16 * 1024 * 1024 }); }
        catch { /* the contract is the judge, not the exit code */ }
      } else {
        await runClaude(`You are one step inside a rabadon pipeline. Do EXACTLY this step, nothing beyond it:\n${instr}\n\nContext: overall task = "${task}". Steps already completed: ${done.map((d) => d.id).join(', ') || 'none'}. Work in the current directory. When the step is done, stop.`, { cwd: dir, worker: true });
      }
    };

    await execStep(step.do);
    let fails = checkContract(step.contract, dir);

    if (fails.length) {
      // THE GATE CAUGHT THE DEVIATION — this is the eclectic moment, killed here.
      emit('CHECK_FAIL', { step: step.id, fails });
      emit('REPAIR_START', { step: step.id, attempt: 1, fixing: fails.map((f) => f.check) });
      log(`  ✗ gate ${step.id}: ${fails[0].why.slice(0, 100)} — pivoting (repair)`);
      await runClaude(`You are rabadon's repair inside a pipeline. Step "${step.id}" (${step.do}) violated its contract:\n${fails.map((f) => `- ${f.check}: ${f.why}`).join('\n')}\nFix the project state so the contract passes. Smallest correct fix. Do not touch anything beyond this step's scope.`, { cwd: dir, worker: true });
      fails = checkContract(step.contract, dir);
      emit(fails.length ? 'REPAIR_FAIL' : 'REPAIR_OK', { step: step.id, attempt: 1 });
    }

    if (fails.length) {
      if (replans >= maxReplans) {
        emit('STOP', { reason: 'CHECK_FAILED', detail: `${step.id} still broken after repair; replan budget spent` });
        emit('RUN_DONE', { verdict: 'CHECK_FAILED' });
        emit.close();
        throw new Error(`rabadon do: gate ${step.id} would not pass and the replan budget is spent — stopped fail-closed, nothing eclectic was produced.`);
      }
      replans++;
      log(`  ⟲ pivot: replanning the remaining pipeline from ${step.id} (${replans}/${maxReplans})`);
      const replan = parseJson(await runClaude(
        PLAN_PROMPT(task, dir, tree) + `\n\nIMPORTANT: steps ${done.map((d) => d.id).join(', ') || '(none)'} are DONE and must not be repeated. Step "${step.id}" failed its contract: ${fails.map((f) => f.why).join('; ')}. Plan ONLY the remaining path to done, taking the failure into account.`,
        { cwd: dir, timeoutMs: 5 * 60000 }));
      if (!replan || !Array.isArray(replan.steps) || !replan.steps.length) throw new Error('rabadon do: replan failed');
      queue = [...replan.steps];
      if (replan.accept) plan.accept = replan.accept;
      continue;
    }

    emit('STEP_OK', { step: step.id });
    done.push(step);
  }

  const acceptFails = checkContract(plan.accept, dir);
  if (acceptFails.length) {
    emit('STOP', { reason: 'CHECK_FAILED', detail: `acceptance: ${acceptFails.map((f) => f.check).join(', ')}` });
    emit('RUN_DONE', { verdict: 'CHECK_FAILED' });
    emit.close();
    throw new Error(`rabadon do: all steps passed their gates but the ACCEPTANCE contract failed: ${acceptFails.map((f) => `${f.check} (${f.why})`).join('; ')} — reporting honest failure, not an eclectic success.`);
  }
  emit('RUN_DONE', { verdict: 'PASS' });
  emit.close();
  log(`accept: PASS — the output is the one that was asked for. (${done.length} steps, ${replans} pivot(s))`);
}

File: native/do.cpp
// rabadon-do — task in, gate-verified output out. C++17, zero deps.
//
// Two moves, both honest about where the LLM sits:
//   PLAN   one claude -p call decomposes the task into a pipeline whose every
//          step carries an un-gameable contract (differential / forbidden where
//          behavior must hold; cmd/fileExists/fileContains otherwise). The plan
//          is a PROPOSAL — it is enforced, not trusted.
//   RUN    exec rabadon-loop, which gates every step with rabadon-verify and
//          repairs-then-continues, bounded, fail-closed.
//
// Remove the LLM and what remains: the loop, the verify arbiter, the bounds,
// the spool. The planner is the one replaceable LLM part; everything that
// decides pass/fail is deterministic C++.
//
// Usage:  rabadon-do "<task>" [dir]

#include <cstdio>
#include <cstdlib>
#include <string>
#include <fstream>
#include <unistd.h>
#include <sys/stat.h>
#include "cli_help.h"

using std::string;

static void write_file(const string& p,const string& s){ std::ofstream f(p,std::ios::binary|std::ios::trunc); f<<s; }

static string run_capture(const string& cmd){
  string out; char buf[8192]; FILE* p=popen(cmd.c_str(),"r"); if(!p) return "";
  size_t n; while((n=fread(buf,1,sizeof buf,p))>0) out.append(buf,n); pclose(p); return out;
}

// feed prompt to `claude -p` on stdin, return its stdout
static string claude(const string& prompt,const string& dir){
  string tmp=dir+"/.rabadon/.plan-prompt.txt"; write_file(tmp,prompt);
  string cmd="cd \""+dir+"\" && claude -p --output-format text < \""+tmp+"\" 2>/dev/null";
  string out=run_capture(cmd); unlink(tmp.c_str()); return out;
}

// strip ``` fences if the model wrapped the JSON
static string unfence(string s){
  size_t a=s.find('{'); size_t b=s.rfind('}');
  if(a!=string::npos && b!=string::npos && b>=a) return s.substr(a,b-a+1);
  return s;
}

static const char* kHelp =
  "rabadon-do — one task in, a gate-verified pipeline out.\n"
  "One model call decomposes the task into steps whose every contract is checked\n"
  "by machine; rabadon-loop then runs it, repairs what fails, and refuses a fix\n"
  "that only pretends to pass.\n"
  "\n"
  "usage: rabadon-do \"<task>\" [dir]\n"
  "\n"
  "  \"<task>\"   what you want done, in one sentence.\n"
  "  [dir]      the project it runs in (default: the current directory).\n"
  "  -h, --help this screen. costs nothing — no model call is made.\n"
  "\n"
  "environment:\n"
  "  RABADON_PROPOSER    the coding-agent command the loop calls for repairs.\n"
  "  RABADON_MAX_REPAIRS repair attempts per step (default 1).\n"
  "\n"
  "example:\n"
  "  rabadon-do \"add a --json flag to the stats command\" ~/src/myrepo\n";

int main(int argc,char** argv){
  // FIRST statement, before the repo walk and before claude() — `rabadon-do
  // --help` used to take the flag as the task and open a paid model call, so
  // the one word a new user types first HUNG and cost money.
  rb_help(argc, argv, kHelp);

  if(argc<2){ fprintf(stderr,"usage: rabadon-do \"<task>\" [dir]\n  run `rabadon-do --help`\n"); return 2; }
  string task=argv[1];
  // a task is a sentence, never a flag. without this, a mistyped option became
  // the prompt and was billed.
  if(rb_is_flag(task.c_str())) rb_unknown_flag("rabadon-do", task.c_str());
  string dir=argc>2?argv[2]:".";
  { char buf[4096]; if(dir=="."){ if(getcwd(buf,sizeof buf)) dir=buf; } }
  mkdir((dir+"/.rabadon").c_str(),0755);

  // real repo structure (depth-limited, not just 40 top names) so the planner
  // can actually see what it must contract over
  string tree=run_capture("cd \""+dir+"\" && (git ls-files 2>/dev/null | head -200 || find . -maxdepth 3 -type f -not -path './.git/*' | head -200)");

  string self=argv[0]; size_t sl=self.rfind('/'); string bindir=sl==string::npos?".":self.substr(0,sl);

  const string prompt=
    "You are rabadon's planner. Decompose the task into a pipeline whose EVERY step is verifiable BY MACHINE. Return ONLY JSON, no prose, no fences:\n"
    "{ \"steps\": [ { \"id\":\"s1\", \"kind\":\"cmd\"|\"work\", \"do\":\"<cmd: literal shell command | work: one narrow instruction for a coding agent>\",\n"
    "  \"contract\":[ <one or more checks> ] } ], \"accept\":[ <task-level definition of done> ] }\n\n"
    "Check types (prefer the un-gameable ones — a repair must not be able to fake them):\n"
    "  {\"type\":\"differential\",\"run\":\"<shell cmd>\",\"expect\":\"<exact stdout that proves correct behavior>\"}  // runs the code, compares behavior; use to PIN behavior\n"
    "  {\"type\":\"forbidden\",\"path\":\"<file>\",\"sha\":\"<leave empty, rabadon fills>\"}  // a file the work must NOT change\n"
    "  {\"type\":\"cmd\",\"run\":\"<shell cmd>\",\"passPattern\":\"<string that appears only on success; omit for exit-0>\"}\n"
    "  {\"type\":\"fileExists\",\"path\":\"<file>\"}   {\"type\":\"fileContains\",\"path\":\"<file>\",\"pattern\":\"<text>\"}\n\n"
    "Laws: 1-6 steps. EVERY step and the accept[] get at least one check. Prefer differential whenever the task has observable behavior (a function's output, a command's result) — that is what makes the gate un-gameable. Use RELATIVE paths in fileExists/fileContains/forbidden and relative commands in run (they execute with the project dir as cwd); do NOT use absolute paths. Steps run in: "+dir+"\n"
    "Repo files:\n"+tree.substr(0,4000)+"\n\nTHE TASK: "+task;

  fprintf(stderr,"rabadon do: planning (one model call)…\n");
  string raw=claude(prompt,dir);
  string plan=unfence(raw);
  if(plan.find("\"steps\"")==string::npos){ fprintf(stderr,"rabadon do: planner returned no usable pipeline.\n%s\n",raw.substr(0,300).c_str()); return 1; }
  string planPath=dir+"/.rabadon/do-plan.json";
  write_file(planPath,plan);
  fprintf(stderr,"rabadon do: plan written to .rabadon/do-plan.json — running under the gate.\n");

  // hand off to the loop: it gates every step with rabadon-verify, repairs, continues
  string loop=bindir+"/rabadon-loop";
  execl(loop.c_str(),loop.c_str(),dir.c_str(),planPath.c_str(),(char*)nullptr);
  // if execl returns, fall back to system()
  string cmd="\""+loop+"\" \""+dir+"\" \""+planPath+"\"";
  return system(cmd.c_str())==0?0:1;
}
